xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go

```diff
@@ -302,3 +313,217 @@ func (s) TestErrorFromParentLB_ResourceNotFound(t *testing.T) {
+}
+
+func newTestChildPolicyBalancerBuilder(b balancer.Builder) *testChildPolicyBalancerBuilder {
+	return &testChildPolicyBalancerBuilder{
+		scStateCh:       make(chan balancer.SubConnState),
+		cfgCh:           make(chan json.RawMessage),
+		priorityBuilder: b,
+	}
+}
+
+func (b *testChildPolicyBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	priorityLB := b.priorityBuilder.Build(cc, opts)
+	return &testChildPolicyBalancer{
+		Balancer:  priorityLB,
+		scStateCh: b.scStateCh,
+	}
+}
+
+func (b *testChildPolicyBalancerBuilder) Name() string {
+	return priority.Name
+}
+
+func (b *testChildPolicyBalancerBuilder) ParseConfig(lbCfg json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	select {
+	case b.cfgCh <- lbCfg:
+	default:
+	}
+	return b.priorityBuilder.(balancer.ConfigParser).ParseConfig(lbCfg)
+}
+
+type testChildPolicyBalancer struct {
+	balancer.Balancer
+	scStateCh chan balancer.SubConnState
+}
+
+func (b *testChildPolicyBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	select {
+	case b.scStateCh <- state:
+	default:
+	}
+
+	b.Balancer.UpdateSubConnState(sc, state)
+}
+
+func (s) TestSubConnStateChangePropagationToChildPolicy(t *testing.T) {
+	// Unregister the priority balancer builder for the duration of this test,
+	// and register a policy under the same name which makes subConn state
+	// changes pushed to it available to the test.
+	priorityBuilder := balancer.Get(priority.Name)
+	internal.BalancerUnregister(priorityBuilder.Name())
+	testChildPolicy := newTestChildPolicyBalancerBuilder(priorityBuilder)
+	balancer.Register(testChildPolicy)
+	defer balancer.Register(priorityBuilder)
+
+	managementServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{})
+	defer cleanup()
+
+	server := stubserver.StartTestService(t, nil)
+	defer server.Stop()
+
+	// Configure cluster and endpoints resources in the management server.
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(clusterName, edsServiceName, e2e.SecurityLevelNone)},
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{e2e.DefaultEndpoint(edsServiceName, "localhost", []uint32{testutils.ParsePort(t, server.Address)})},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create xDS client, configure cds_experimental LB policy with a manual
+	// resolver, and dial the test backends.
+	cc, cleanup := setupAndDial(t, bootstrapContents)
+	defer cleanup()
+
+	// Spawn a goroutine which ensures that subConn state changes are propagated
+	// to the child policy. We *only* wait for READY state since we expect to
+	// get to that state when the below RPC succeeds.
+	doneCh := make(chan error, 1)
```

Comment:
I've always done this pattern with testutils.NewChannel. The problem is the extra typecast after receiving. I like this way better, I'll do this from now on (verifying events in a separate goroutine).